paasta_tools/cleanup_kubernetes_jobs.py

```diff
@@ -0,0 +1,189 @@
+#!/usr/bin/env python
+# Copyright 2019-2020 Yelp Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Usage: ./cleanup_kubernetes_jobs.py [options]
+
+Clean up kubernetes apps that aren't supposed to run on this cluster by deleting them.
+
+Gets the current app list from kubernetes, and then a 'valid_app_list'
+via utils.get_services_for_cluster
+
+If an app in the kubernetes app list isn't in the valid_app_list, it's
+deleted.
+
+Command line options:
+
+- -d <SOA_DIR>, --soa-dir <SOA_DIR>: Specify a SOA config dir to read from
+- -v, --verbose: Verbose output
+- -t <KILL_THRESHOLD>, --kill-threshold: The decimal fraction of apps we think
+    is sane to kill when this job runs
+- -f, --force: Force the killing of apps if we breach the threshold
+"""
+import argparse
+import logging
+import sys
+import traceback
+from contextlib import contextmanager
+from typing import Generator
+
+from kubernetes.client import V1Deployment
+from kubernetes.client import V1StatefulSet
+
+from paasta_tools.kubernetes.applicaton.tools import Application
+from paasta_tools.kubernetes.applicaton.tools import list_namespaced_applications
+from paasta_tools.kubernetes_tools import KubeClient
+from paasta_tools.utils import _log
+from paasta_tools.utils import DEFAULT_SOA_DIR
+from paasta_tools.utils import get_services_for_cluster
+from paasta_tools.utils import load_system_paasta_config
+# import pysensu_yelp
+# from paasta_tools.monitoring_tools import send_event
+# import pysensu_yelp
+
+
+log = logging.getLogger(__name__)
+APPLICATION_TYPES = [V1StatefulSet, V1Deployment]
+
+
+class DontKillEverythingError(Exception):
+    pass
+
+
+@contextmanager
+def notify(application: Application, soa_dir: str) -> Generator:
```

Comment:
Not sure about the naming? If this is to notify of other things then we need to pass the "message" as an argument?

Reply:
This function is supposed to send_events to sensu and log this event. I was expecting that we have sensu checks like https://github.com/Yelp/paasta/blob/master/paasta_tools/cleanup_marathon_jobs.py#L84 . It's probably a bad name. I guess I can change it to alert_state_change? Although we don't have sensu checks right now, I think we should add in the future. I suggest we just rename it and leave it here for now?